src/wasm/wasm-emscripten.cpp

```diff
@@ -445,9 +445,40 @@ void EmscriptenGlueGenerator::replaceStackPointerGlobal() {
 }
 
 std::vector<Address> getSegmentOffsets(Module& wasm) {
+  std::unordered_map<unsigned, Address> passiveOffsets;
+  if (wasm.features.hasBulkMemory()) {
+    // Fetch passive segment offsets out of memory.init instructions
+    struct OffsetSearcher : PostWalker<OffsetSearcher> {
+      std::unordered_map<unsigned, Address>& offsets;
+      OffsetSearcher(std::unordered_map<unsigned, Address>& offsets)
+        : offsets(offsets) {}
+      void visitMemoryInit(MemoryInit* curr) {
+        auto* dest = curr->dest->dynCast<Const>();
+        if (!dest) {
+          return;
+        }
+        auto it = offsets.find(curr->segment);
+        if (it != offsets.end()) {
+          Fatal() << "Cannot get offset of passive segment initialized "
+                     "multiple times";
+        }
+        offsets[curr->segment] = dest->value.geti32();
+      }
+    } searcher(passiveOffsets);
+    searcher.walkModule(&wasm);
+  }
   std::vector<Address> segmentOffsets;
   for (unsigned i = 0; i < wasm.memory.segments.size(); ++i) {
-    if (auto* addrConst = wasm.memory.segments[i].offset->dynCast<Const>()) {
+    auto& segment = wasm.memory.segments[i];
+    if (segment.isPassive) {
+      auto it = passiveOffsets.find(i);
+      if (it != passiveOffsets.end()) {
+        segmentOffsets.push_back(it->second);
+      } else {
+        // This was a non-constant offset (perhaps TLS)
```

Comment:
I don't understand what this does. What will happen with that 0?

Reply:
Basically if AsmConstWalker or EmJsWalker try to extract anything from such a segment they will silently get a bogus address. I added a condition to `stringAtAddr`, which seems to be the ultimate consumer of these segment offsets, to make it skip any segment with a 0 offset, which can never be valid.

Reply:
Hmm, why can't 0 be a valid offset? Doesn't that mean it starts at address 0 which is valid?

Reply:
0 can't be a valid segment offset because that would mean the linker tried to put some data at address 0, which is of course inaccessible from C.